src/baseline.rs

```diff
@@ -340,3 +356,82 @@ fn need_retry(res: Result<(), crates_index::Error>) -> anyhow::Result<bool> {
         Err(err) => Err(err.into()),
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crates_index::{Crate, Version};
+
+    use super::choose_baseline_version;
+
+    fn new_mock_version(version_name: &str, yanked: bool) -> Version {
+        // `crates_index::Version` cannot be created explicitly, as all its fields
+        // are private, so we use the fact that it can be deserialized.
+        serde_json::from_value(serde_json::json!({
+            "name": "test-crate",
+            "vers": version_name,
+            "deps": [],
+            "features": {},
+            "yanked": yanked,
+            "cksum": "00".repeat(32),
+        }))
+        .unwrap()
+    }
+
+    fn new_crate(versions: &[Version]) -> Crate {
+        // `crates_index::Crate` cannot be created explicitly, as its field
+        // is private, so we use the fact that it can be deserialized.
+        serde_json::from_value(serde_json::json!({ "versions": versions })).unwrap()
+    }
+
+    #[test]
+    fn choose_baseline_version_yanked() {
+        let crate_ = new_crate(&[
+            new_mock_version("1.2.0", false),
+            new_mock_version("1.2.1", true),
+        ]);
+        let current_version = semver::Version::new(1, 2, 2);
+        assert_eq!(
+            choose_baseline_version(&crate_, Some(&current_version)).unwrap(),
+            "1.2.0".to_string()
+        );
+    }
+
+    #[test]
+    fn choose_baseline_version_not_latest() {
+        let crate_ = new_crate(&[
+            new_mock_version("1.2.0", false),
+            new_mock_version("1.2.1", false),
+        ]);
+        let current_version = semver::Version::new(1, 2, 0);
+        assert_eq!(
+            choose_baseline_version(&crate_, Some(&current_version)).unwrap(),
+            "1.2.0".to_string()
+        );
+    }
+
+    #[test]
+    fn choose_baseline_version_pre_release() {
+        let crate_ = new_crate(&[
+            new_mock_version("1.2.0", false),
+            new_mock_version("1.2.1-rc1", false),
+        ]);
+        let current_version = semver::Version::parse("1.2.1-rc2").unwrap();
+        assert_eq!(
+            choose_baseline_version(&crate_, Some(&current_version)).unwrap(),
+            "1.2.0".to_string()
+        );
+    }
+
+    #[test]
+    fn choose_baseline_version_no_current() {
```

Comment:
This function name is possibly misleading. From the name, I would expect the test to have a `current` version and a list of crates that doesn't have this `current` version.

Reply:
And same question here, do you have some idea how can we name it better?

Reply:
What do you think about `baseline_choosing_logic_without_current_picks_latest_normal`? A bit long, but I don't see any reason to make it shorter.